### PR TITLE
Modify camera all

### PR DESCRIPTION
**Pull request by ligenxxxx**
a lot of code is modified.
- optimized camera menu and config.
- add shutter for HDZ microV2/nano 90 camera. still need to modify.
- add led mode for all HDZ camera but didn't test.
- remove wb manual mode number.
- support 3 profiles for all camera.
- if vtx's camera type is changed, vtx will reset all camera setting.
- I can't remember what else.

### CODE REVIEW

File: src/camera.c [hunk 1]
```diff
@@ -109,1237 +100,571 @@ uint8_t CamDetect() {
...
-void camera_check_and_save_parameters() {
-    uint8_t i = 0;
-    uint8_t j = 0;
-
-    if (cameraID == 0)
-        return;
-
-    if (camProfile_EEP >= Profile_Max)
-        camProfile_EEP = Profile_MicroV2;
+void camMenuDrawValue(void) {
```
Comment (bkleiner) [inline src/camera.c:296]: naming: `camera_menu_draw_value`

File: src/camera.c [hunk 2]
```diff
@@ -109,1237 +100,571 @@ uint8_t CamDetect() {
...
-    RUNCAM_Write(cameraID, 0x000694, 0x00000311);
-#ifdef _DEBUG_CAMERA
-    debugf("\r\nRUNCAM NightMode:%02x", (uint16_t)val);
-#endif
 }
 
-void Runcam_SetWB(uint8_t *wbRed, uint8_t *wbBlue, uint8_t wbMode) {
-    uint32_t wbRed_u32 = 0x02000000;
-    uint32_t wbBlue_u32 = 0x00000000;
+void camMenuDrawBracket(void) {
```
Comment (bkleiner) [inline src/camera.c:288]: naming: `camera_menu_draw_bracket`

File: src/camera.c [hunk 3]
```diff
@@ -109,1237 +100,571 @@ uint8_t CamDetect() {
...
-            camCfg_EEP[i].contrast = camera_read_eep_parameter((i << 4) + EEP_ADDR_CAM_CONTRAST);
-            camCfg_EEP[i].hvFlip = camera_read_eep_parameter((i << 4) + EEP_ADDR_CAM_HVFLIP);
-            camCfg_EEP[i].nightMode = camera_read_eep_parameter((i << 4) + EEP_ADDR_CAM_NIGHTMODE);
-            camCfg_EEP[i].wbMode = camera_read_eep_parameter((i << 4) + EEP_ADDR_CAM_WBMODE);
-            for (j = 0; j < WBMODE_MAX; j++) {
-                camCfg_EEP[i].wbRed[j] = camera_read_eep_parameter((i << 4) + j + EEP_ADDR_CAM_WBRED);
-                camCfg_EEP[i].wbBlue[j] = camera_read_eep_parameter((i << 4) + j + EEP_ADDR_CAM_WBBLUE);
-            }
-        }
+void camMenuInit(void) {
```
Comment (bkleiner) [inline src/camera.c:390]: naming: `camera_menu_init`

File: src/camera.h [hunk 4]
```diff
@@ -9,40 +9,32 @@
...
-#define SATURATION_MAX 0x06
-#define CONTRAST_MIN   0x00
-#define CONTRAST_MAX   0x02
-#define HVFLIP_MAX     0x01
-#define NIGHTMODE_MAX  0x01
-#define WBMODE_MIN     0x00
-#define WBMODE_MAX     0x04
+typedef enum {
+    CAMERA_TYPE_UNKNOW,
+    CAMERA_TYPE_RUNCAM_MICRO_V1,
```
Comment (bkleiner) [inline src/camera.h:24]: merge this and the defines? eg `CAMERA_TYPE_RUNCAM_MICRO_V1 = 0x42`

File: src/i2c_device.c [hunk 5]
```diff
@@ -122,15 +117,18 @@ void Init_TC3587(uint8_t fmt) {
...
     I2C_Write16(ADDR_TC3587, 0x000c, 0x0101);     // mclk
-    I2C_Write16(ADDR_TC3587, 0x0018, 0x0111);     // pll //0111
-    I2C_Write16(ADDR_TC3587, 0x0018, 0x0113);     // pll //0113
+    I2C_Write16(ADDR_TC3587, 0x0018, 0x0111);     // pll
+    I2C_Write16(ADDR_TC3587, 0x0018, 0x0113);     // pll
     I2C_Write16(ADDR_TC3587, 0x0002, 0x0001);
     I2C_Write16(ADDR_TC3587, 0x0002, 0x0000);
     // WAIT(100);
+#ifdef _DEBUG_TC3587
+    debugf("\r\Init TC3587 done");
```
Comment (bkleiner) [inline src/i2c_device.c:132]: missing `\n`

File: src/camera.h [hunk 6]
```diff
@@ -60,48 +52,44 @@ typedef enum {
...
+
+    CAM_STATUS_END,
 } camStatusType_e;
 
 void CameraInit();
 uint8_t camStatusUpdate(uint8_t op);
-void camMenuInit();
+void camMenuInit(void);
 
 extern uint8_t camRatio;
```
Comment (bkleiner) [inline src/camera.h:89]: is this still used?

File: src/camera.c [hunk 7]
```diff
@@ -1,105 +1,96 @@
...
+uint8_t camera_setting_reg_menu[CAMERA_SETTING_NUM];
+uint8_t camera_profile_menu;
+
+uint8_t video_format = VDO_FMT_720P60;
 uint8_t camRatio = 0;
 uint8_t camMenuStatus = CAM_STATUS_IDLE;
-uint8_t bw_changed = 0;
 
 void camMenuStringUpdate(uint8_t status);
 void camMenuSetVdoRatioInit();
```
Comment (bkleiner) [inline src/camera.c:26]: these two appear unused

File: src/camera.c [hunk 8]
```diff
@@ -109,1237 +100,571 @@ uint8_t CamDetect() {
...
-        1: night mode on
-    */
-    if (cameraID != RUNCAM_MICRO_V2 || cameraID != RUNCAM_MICRO_V3)
-        return;
+void Cam_Button_ENTER() { WriteReg(0, 0x14, 0x32); }
+void Cam_Button_RIGHT() { WriteReg(0, 0x14, 0x58); }
+void Cam_Button_DOWN() { WriteReg(0, 0x14, 0x64); }
+void Cam_Button_LEFT() { WriteReg(0, 0x14, 0x3F); }
+void Cam_Button_UP() { WriteReg(0, 0x14, 0x4B); }
+void Cam_Button_MID() { WriteReg(0, 0x14, 0x00); }
```
Comment (bkleiner) [inline src/camera.c:261]: naming: `camera_button_*`
also convert to macro?

File: src/camera.c [hunk 9]
```diff
@@ -109,1237 +100,571 @@ uint8_t CamDetect() {
...
     return ret;
 }
-#endif
-
-void Runcam_SetHVFlip(uint8_t val) {
-    if (cameraID != RUNCAM_MICRO_V2 || cameraID == RUNCAM_MICRO_V3)
-        return;
 
-    camCfg_Cur.hvFlip = val;
+void CameraInit(void) {
```
Comment (bkleiner) [inline src/camera.c:246]: naming: `camera_init`

File: src/camera.c [hunk 10]
```diff
@@ -109,1237 +100,571 @@ uint8_t CamDetect() {
...
 
-void Cam_Button_ENTER() { WriteReg(0, 0x14, 0x32); }
-void Cam_Button_RIGHT() { WriteReg(0, 0x14, 0x58); }
-void Cam_Button_DOWN() { WriteReg(0, 0x14, 0x64); }
-void Cam_Button_LEFT() { WriteReg(0, 0x14, 0x3F); }
-void Cam_Button_UP() { WriteReg(0, 0x14, 0x4B); }
-void Cam_Button_MID() { WriteReg(0, 0x14, 0x00); }
-
 #ifdef USE_MSP
 uint8_t camStatusUpdate(uint8_t op) {
```
Comment (bkleiner) [inline src/camera.c:545]: naming: `camera_status_update`

File: src/hardware.c [hunk 11]
```diff
@@ -190,7 +190,7 @@ void Setting_Save() {
     uint8_t rcv = 0;
 
 #if (0)
-    if (cameraID == 0) {
+    if (cameraType == 0) {
```
Comment (bkleiner) [inline src/hardware.c:193]: should be `camera_type`

File: src/msp_displayport.c [hunk 12]
```diff
@@ -89,7 +89,8 @@ uint8_t osd_menu_offset = 0;
 #ifdef USE_MSP
 
 uint8_t msp_cmp_fc_variant(const char *variant) {
-    for (uint8_t i = 0; i < sizeof(fc_variant); i++) {
```
Comment (bkleiner) [inline src/msp_displayport.c:92]: :/ *oops*

File: src/camera.h [hunk 13]
```diff
@@ -60,48 +52,44 @@ typedef enum {
...
     CAM_STATUS_SAVE_EXIT,
-    CAM_STATUS_SET_VDO_RATIO_4_3,
-    CAM_STATUS_SET_VDO_RATIO_16_9_CROP,
-    CAM_STATUS_SET_VDO_RATIO_16_9_FULL,
-    CAM_STATUS_SET_VDO_RATIO_RETURN,
-    CAM_STATUS_END,
     CAM_STATUS_REPOWER,
+
+    CAM_STATUS_END,
 } camStatusType_e;
```
Comment (bkleiner) [inline src/camera.h:83]: naming: `camera_status_e`

File: src/camera.h [hunk 14]
```diff
@@ -60,48 +52,44 @@ typedef enum {
...
-    uint8_t wbMode;
-    uint8_t wbRed[WBMODE_MAX];
-    uint8_t wbBlue[WBMODE_MAX];
-} cameraConfig_t;
+typedef enum {
+    item_enable,
+    item_min,
+    item_max,
+    item_default,
+} camera_setting_attribute;
```
Comment (bkleiner) [inline src/camera.h:60]: naming: missing `_e` suffix

File: src/camera.h [hunk 15]
```diff
@@ -60,48 +52,44 @@ typedef enum {
...
-    uint8_t saturation;
-    uint8_t contrast;
-    uint8_t hvFlip;
-    uint8_t nightMode;
-    uint8_t wbMode;
-    uint8_t wbRed[WBMODE_MAX];
-    uint8_t wbBlue[WBMODE_MAX];
-} cameraConfig_t;
+typedef enum {
+    item_enable,
```
Comment (bkleiner) [inline src/camera.h:56]: Most other enums are UPPER, lets keep it consistent?

**Review by bkleiner — COMMENTED** [review]
So much better, both in terms of code readability and flash usage.
Removes about 4K on sdcc (42977 vs 46281, whoop vtx).
Mostly nit-picks on the review.

File: src/camera.h (hunk 4 above)
Comment (ligenxxxx) [reply src/camera.h:24]: It seems so, but it is not the case. A device address may correspond to multiple camera types, which can be distinguished by reading a certain register, but the camera cannot be implemented temporarily.

File: src/camera.h (hunk 6 above)
Comment (ligenxxxx) [reply src/camera.h:89]: Yes.

**Review by bkleiner — COMMENTED** [review]
- two minor code nitty-picks
- as already noted on discord, settings dont work on v1 cams
- shutter speed settings is nice
- i vote for Video Mode vs Aspect Ratio

**Pull request merged**